chore: Migrate from Poetry to UV, added health checkpoint, and improved dockerimage

spa5k opened the pull request:
- Add .dockerignore to optimize Docker build context
- Refactor docker-compose files for CPU and GPU modes to add health check.
- Update Dockerfile with multi-stage build and improved caching
- Add Makefile for simplified project management
- Implement health check endpoint in document converter route
- Optimize container runtime and model download process
- Improve documentation and setup instructions

spa5k committed:
chore: Migrate from Poetry to UV and update project configuration

spa5k committed:
feat: Enhance Docker and deployment configuration

- Add .dockerignore to optimize Docker build context
- Refactor docker-compose files for CPU and GPU modes
- Update Dockerfile with multi-stage build and improved caching
- Add Makefile for simplified project management
- Implement health check endpoint in document converter route
- Optimize container runtime and model download process
- Improve documentation and setup instructions

spa5k committed:
Merge remote-tracking branch 'origin/main' into docker-uv

spa5k commented:
@drmingler do you mind taking a look?

drmingler commented:
> @drmingler do you mind taking a look?

@spa5k Awesome work! Thank you. I will have a look and get back. Meanwhile can we have a unified lint rule so that we can avoid modifying the code formatting?

spa5k commented:
Sure, I tried not to change much formatting, will reset wherever it happened. UV also provides a formatter, maybe it can be used

spa5k commented:
What about this? Let's merge this, then I will create a PR to implement [Ruff](https://docs.astral.sh/ruff/), as a GitHub action, and a pre commit in next PR, otherwise, there will be too many changes in this PR.

@drmingler

drmingler commented:
> What about this? Let's merge this, then I will create a PR to implement [Ruff](https://docs.astral.sh/ruff/), as a GitHub action, and a pre commit in next PR, otherwise, there will be too many changes in this PR.
> 
> @drmingler

Okay that makes sense. I will merge this over the weekend. Thank you

spa5k committed:
refactor: Improve document conversion route and schema with enhanced type safety and API documentation

- Add comprehensive response examples and status codes
- Introduce JobStatus and ImageType enums for better type safety
- Enhance route error handling and status reporting
- Add detailed field descriptions and validation
- Improve health check endpoint with more granular service status reporting

spa5k committed:
refactor: Simplify document conversion schemas and routes and improved docker caching

spa5k commented:
Thanks, the PR is now in quite good stage, I tested the caching, and schema changes, including generating SDK using it for Go, and everything seems to be working well.

drmingler commented on README.md:
````diff
@@ -50,20 +50,20 @@
+### 1. Install uv (if not already installed)
 ```bash
-curl -sSL https://install.python-poetry.org | python3 -
+curl -LsSf https://astral.sh/uv/install.sh | sh
````
Can we also add instructions on how to install uv for those using windows or add a link to the uv installation doc? @spa5k

drmingler commented on README.md:
````diff
@@ -144,16 +144,43 @@ REDIS_HOST=redis://redis:6379/0
-docker-compose -f docker-compose.cpu.yml up --build --scale celery_worker=1
+# Build and run in GPU mode with 1 worker
+make docker-build-gpu
+make docker-run-gpu
+
+# Or build and run with multiple workers
+make docker-run-gpu WORKER_COUNT=3
 ```
````
Should be "docker-start-cpu" not "docker-run-gpu"

drmingler reviewed (changes requested):
<img width="1454" alt="Screenshot 2025-03-08 at 18 59 04" src="https://github.com/user-attachments/assets/78864be5-a203-41cc-a120-154010c6b657" />
It doesn't work on GPU but works on CPU. Modules are not installed during docker build.

spa5k committed:
refactor: Optimize Dockerfile and Docker configurations for improved build and runtime performance

- Simplify Dockerfile with single-stage build
- Improve PyTorch and EasyOCR model installation based on system architecture
- Update docker-compose files to remove runtime target
- Add auto-detection for CPU/GPU docker start in Makefile
- Enhance .dockerignore with additional Python-related exclusions
- Add entrypoint script for consistent container startup

spa5k committed:
feat: Add GPU detection script and update Docker configurations

- Introduce detect_gpu.sh script for dynamic GPU and Docker configuration detection
- Update docker-compose.cpu.yml and docker-compose.gpu.yml to use uv for running commands
- Enhance Dockerfile with multi-stage build, improved GPU/CPU detection, and model downloading
- Add NVIDIA GPU capabilities and platform configuration in Docker Compose files
- Improve runtime and build performance with better caching and environment setup

spa5k committed:
docs: Update README.md Docker commands to use docker-start instead of docker-run

spa5k commented:
Can you check the GPU part now? If it is not working, please move the pytorch and model downloads part to the runtime stage

drmingler commented on document_converter/route.py:
```diff
@@ -65,54 +104,120 @@ async def convert_multiple_documents(
 )
 async def get_conversion_job_status(job_id: str):
-    return document_converter_service.get_single_document_task_result(job_id)
+    try:
+        result = document_converter_service.get_single_document_task_result(job_id)
+        
+        # Return 202 Accepted if job is still in progress
+        if result.status in ["IN_PROGRESS"]:
+            return JSONResponse(
+                status_code=status.HTTP_202_ACCEPTED,
+                content=result.dict(exclude_none=True)
+            )
+            
+        # Return 422 for failed jobs
+        if result.status == "FAILURE":
+            return JSONResponse(
+                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+                content=result.dict(exclude_none=True)
+            )
+            
+        # Return 200 OK for successful jobs
+        return JSONResponse(
+            status_code=status.HTTP_200_OK,
+            content=result.dict(exclude_none=True)
+        )
+    except KeyError:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail=f"Job not found: {job_id}"
+        )
```
Please take this off, during job processing we do not raise errors, we return a status of FAILURE if the job failed and the reason for the failure in the error field. It was intentionally done that way.

drmingler commented on document_converter/route.py:
```diff
@@ -121,15 +226,90 @@ async def create_batch_conversion_job(
-    """Get the status and results of a batch conversion job."""
-    return document_converter_service.get_batch_conversion_task_result(job_id)
+    try:
+        result = document_converter_service.get_batch_conversion_task_result(job_id)
+        
+        # Return 202 Accepted if the batch job or any sub-job is still in progress
+        if result.status in ["IN_PROGRESS"] or any(
+            job.status in ["IN_PROGRESS"]
+            for job in result.conversion_results
+        ):
+            return JSONResponse(
+                status_code=status.HTTP_202_ACCEPTED,
+                content=result.dict(exclude_none=True)
+            )
+            
+        # Return 422 for failed batch jobs
+        if result.status == "FAILURE" or any(
+            job.status == "FAILURE"
+            for job in result.conversion_results
+        ):
+            return JSONResponse(
+                status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
+                content=result.dict(exclude_none=True)
+            )
+            
+        # Return 200 OK for successful batch jobs (all success)
+        return JSONResponse(
+            status_code=status.HTTP_200_OK,
+            content=result.dict(exclude_none=True)
+        )
+    except KeyError:
+        raise HTTPException(
+            status_code=status.HTTP_404_NOT_FOUND,
+            detail=f"Batch job not found: {job_id}"
+        )
```
Same here

drmingler commented:
> Can you check the GPU part now? If it is not working, please move the pytorch and model downloads part to the runtime stage

I will test on gpu and get back to you. @spa5k
